File: scripts/generate-docx.py
#!/usr/bin/env python3
"""
Script para convertir TEAM-05-FULL-DOCUMENTATION.md a .docx
Genera el archivo en: C:/Users/saule/OneDrive/Documentos/Programación Web/
"""

import re
import os
from docx import Document
from docx.shared import Inches, Pt, Cm, RGBColor
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.enum.table import WD_TABLE_ALIGNMENT
from docx.oxml.ns import qn
from docx.oxml import OxmlElement

def read_md_file(path):
    with open(path, 'r', encoding='utf-8') as f:
        return f.read()

def add_code_block(doc, code):
    """Add a formatted code block."""
    p = doc.add_paragraph()
    p.style = doc.styles['Normal']
    # Set background to light gray
    pf = p.paragraph_format
    pf.space_before = Pt(6)
    pf.space_after = Pt(6)
    # Add the code with monospace font
    run = p.add_run(code)
    run.font.name = 'Consolas'
    run.font.size = Pt(9)
    run.font.color.rgb = RGBColor(0x1A, 0x1A, 0x2E)
    # Add shading
    shading = OxmlElement('w:shd')
    shading.set(qn('w:fill'), 'F5F5F5')
    shading.set(qn('w:val'), 'clear')
    p.paragraph_format.element.get_or_add_pPr().append(shading)

def add_table_row(table, cells_data, bold=False, header=False):
    """Add a row to the table."""
    row = table.add_row()
    for i, text in enumerate(cells_data):
        cell = row.cells[i]
        cell.text = ''
        p = cell.paragraphs[0]
        run = p.add_run(str(text))
        run.font.size = Pt(9)
        if bold or header:
            run.bold = True
        if header:
            run.font.color.rgb = RGBColor(0xFF, 0xFF, 0xFF)
            # Add dark header background
            shading = OxmlElement('w:shd')
            shading.set(qn('w:fill'), '2B579A')
            shading.set(qn('w:val'), 'clear')
            cell.paragraphs[0].paragraph_format.element.get_or_add_pPr().append(shading)

def convert_md_to_docx(md_content, output_path):
    doc = Document()
    
    # Set default font
    style = doc.styles['Normal']
    font = style.font
    font.name = 'Calibri'
    font.size = Pt(11)
    
    # Configure heading styles
    for i in range(1, 4):
        heading_style = doc.styles[f'Heading {i}']
        heading_style.font.color.rgb = RGBColor(0x2B, 0x57, 0x9A)
        heading_style.font.bold = True
    
    doc.styles['Heading 1'].font.size = Pt(22)
    doc.styles['Heading 2'].font.size = Pt(16)
    doc.styles['Heading 3'].font.size = Pt(13)
    
    lines = md_content.split('\n')
    in_code_block = False
    code_buffer = []
    in_table = False
    in_list = False
    table_buffer = []
    
    i = 0
    while i < len(lines):
        line = lines[i]
        
        # Handle code blocks
        if line.strip().startswith('```'):
            if in_code_block:
                add_code_block(doc, '\n'.join(code_buffer))
                code_buffer = []
                in_code_block = False
            else:
                in_code_block = True
            i += 1
            continue
        
        if in_code_block:
            code_buffer.append(line)
            i += 1
            continue
        
        # Handle tables
        if line.strip().startswith('|') and line.strip().endswith('|'):
            table_buffer.append(line)
            in_table = True
            i += 1
            continue
        else:
            if in_table and len(table_buffer) >= 3:
                # Process table
                # Parse header
                header_cells = [c.strip() for c in table_buffer[0].strip().split('|') if c.strip()]
                # Check if second line is separator
                if '---' in table_buffer[1]:
                    data_start = 2
                else:
                    data_start = 1
                
                # Create table
                num_cols = len(header_cells)
                table = doc.add_table(rows=1, cols=num_cols)
                table.style = 'Light Grid Accent 1'
                table.alignment = WD_TABLE_ALIGNMENT.CENTER
                
                # Add header row
                for j, header in enumerate(header_cells):
                    cell = table.rows[0].cells[j]
                    cell.text = ''
                    p = cell.paragraphs[0]
                    run = p.add_run(header)
                    run.bold = True
                    run.font.size = Pt(9)
                
                # Add data rows
                for row_idx in range(data_start, len(table_buffer)):
                    row_cells = [c.strip() for c in table_buffer[row_idx].strip().split('|') if c.strip()]
                    if len(row_cells) != num_cols:
                        continue
                    row = table.add_row()
                    for j, cell_text in enumerate(row_cells):
                        cell = row.cells[j]
                        cell.text = ''
                        p = cell.paragraphs[0]
                        run = p.add_run(cell_text)
                        run.font.size = Pt(9)
                
                doc.add_paragraph()  # spacing
                table_buffer = []
                in_table = False
            elif in_table:
                table_buffer = []
                in_table = False
        
        # Handle headings
        if line.startswith('# '):
            doc.add_heading(line[2:].strip(), level=1)
        elif line.startswith('## '):
            doc.add_heading(line[3:].strip(), level=2)
        elif line.startswith('### '):
            doc.add_heading(line[4:].strip(), level=3)
        elif line.startswith('#### '):
            doc.add_heading(line[5:].strip(), level=4)
        elif line.startswith('- ') or line.startswith('* '):
            # Handle list items
            text = line[2:].strip()
            # Check for bold markers **...**
            if '**' in text:
                p = doc.add_paragraph(style='List Bullet')
                parts = re.split(r'(\*\*.*?\*\*)', text)
                for part in parts:
                    if part.startswith('**') and part.endswith('**'):
                        run = p.add_run(part[2:-2])
                        run.bold = True
                    else:
                        p.add_run(part)
            else:
                doc.add_paragraph(text, style='List Bullet')
        elif line.strip() == '---':
            # Horizontal rule
            p = doc.add_paragraph()
            run = p.add_run('_' * 60)
            run.font.color.rgb = RGBColor(0xCC, 0xCC, 0xCC)
            run.font.size = Pt(8)
        elif line.strip() == '':
            if not in_list:
                pass  # Skip empty lines
            in_list = False
        else:
            # Normal paragraph text
            text = line.strip()
            if text:
                # Check for bold markers
                if '**' in text:
                    p = doc.add_paragraph()
                    parts = re.split(r'(\*\*.*?\*\*)', text)
                    for part in parts:
                        if part.startswith('**') and part.endswith('**'):
                            run = p.add_run(part[2:-2])
                            run.bold = True
                        else:
                            p.add_run(part)
                else:
                    p = doc.add_paragraph(text)
                in_list = False
        
        i += 1
    
    # Handle any remaining code block
    if in_code_block and code_buffer:
        add_code_block(doc, '\n'.join(code_buffer))
    
    # Save
    os.makedirs(os.path.dirname(output_path), exist_ok=True)
    doc.save(output_path)
    print(f"✅ Documento DOCX generado en: {output_path}")
    print(f"   Tamaño: {os.path.getsize(output_path) / 1024:.1f} KB")

if __name__ == '__main__':
    md_path = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'docs', 'TEAM-05-FULL-DOCUMENTATION.md')
    output_path = r'C:\Users\saule\OneDrive\Documentos\Programación Web\TEAM-05-TurboPapus-Documentacion-Completa.docx'
    
    if not os.path.exists(md_path):
        print(f"❌ No se encontró el archivo .md: {md_path}")
        exit(1)
    
    md_content = read_md_file(md_path)
    convert_md_to_docx(md_content, output_path)

File: scripts/generate-docx.mjs
#!/usr/bin/env node
/**
 * Script que convierte TEAM-05-FULL-DOCUMENTATION.md a un documento .docx
 * con formato profesional (tablas, código, headings, listas).
 * 
 * Salida: C:\Users\saule\OneDrive\Documentos\Programación Web\TEAM-05-TurboPapus-Documentacion-Completa.docx
 */

import { Document, Packer, Paragraph, TextRun, HeadingLevel, 
         Table, TableRow, TableCell, WidthType, AlignmentType, 
         BorderStyle, ShadingType, convertInchesToTwip } from 'docx';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const MD_PATH = path.resolve(__dirname, '..', 'docs', 'TEAM-05-FULL-DOCUMENTATION.md');
const OUTPUT_PATH = 'C:\\Users\\saule\\OneDrive\\Documentos\\Programación Web\\TEAM-05-TurboPapus-Documentacion-Completa.docx';

// ─── Colores corporativos ─────────────────────────────────────
const COLORS = {
  primary: '2B579A',
  secondary: '1A1A2E',
  codeBg: 'F5F5F5',
  tableHeader: '2B579A',
  tableHeaderText: 'FFFFFF',
  text: '333333',
  link: '0066CC',
};

// ─── Parse helpers ────────────────────────────────────────────
function parseMarkdown(md) {
  const lines = md.split('\n');
  const blocks = [];
  let i = 0;
  
  while (i < lines.length) {
    const line = lines[i];
    
    // Code block
    if (line.trim().startsWith('```')) {
      const lang = line.trim().slice(3).trim();
      const codeLines = [];
      i++;
      while (i < lines.length && !lines[i].trim().startsWith('```')) {
        codeLines.push(lines[i]);
        i++;
      }
      blocks.push({ type: 'code', lang, content: codeLines.join('\n') });
      i++;
      continue;
    }
    
    // Heading
    if (line.startsWith('# ')) {
      blocks.push({ type: 'heading', level: 1, content: line.slice(2).trim() });
      i++;
      continue;
    }
    if (line.startsWith('## ')) {
      blocks.push({ type: 'heading', level: 2, content: line.slice(3).trim() });
      i++;
      continue;
    }
    if (line.startsWith('### ')) {
      blocks.push({ type: 'heading', level: 3, content: line.slice(4).trim() });
      i++;
      continue;
    }
    if (line.startsWith('#### ')) {
      blocks.push({ type: 'heading', level: 4, content: line.slice(5).trim() });
      i++;
      continue;
    }
    
    // Table
    if (line.trim().startsWith('|') && line.trim().endsWith('|')) {
      const tableLines = [];
      while (i < lines.length && lines[i].trim().startsWith('|') && lines[i].trim().endsWith('|')) {
        tableLines.push(lines[i]);
        i++;
      }
      blocks.push({ type: 'table', lines: tableLines });
      continue;
    }
    
    // Horizontal rule
    if (line.trim() === '---') {
      blocks.push({ type: 'hr' });
      i++;
      continue;
    }
    
    // Empty line
    if (line.trim() === '') {
      blocks.push({ type: 'spacer' });
      i++;
      continue;
    }
    
    // List item
    if (line.trim().startsWith('- ') || line.trim().startsWith('* ')) {
      const listItems = [];
      while (i < lines.length && (lines[i].trim().startsWith('- ') || lines[i].trim().startsWith('* '))) {
        listItems.push(lines[i].trim().slice(2).trim());
        i++;
      }
      blocks.push({ type: 'list', items: listItems });
      continue;
    }
    
    // Paragraph (with possible bold markers)
    if (line.trim()) {
      blocks.push({ type: 'paragraph', content: line.trim() });
      i++;
      continue;
    }
    
    i++;
  }
  
  return blocks;
}

function parseInlineFormatting(text) {
  // Split by **bold** markers
  const parts = text.split(/(\*\*.*?\*\*)/);
  return parts.map(part => {
    if (part.startsWith('**') && part.endsWith('**')) {
      return { text: part.slice(2, -2), bold: true };
    }
    return { text: part, bold: false };
  });
}

function parseTable(lines) {
  if (lines.length < 2) return null;
  
  const rows = lines.map(l => 
    l.split('|')
      .slice(1, -1)
      .map(c => c.trim())
  );
  
  // Filter out separator row
  const dataRows = rows.filter(r => !r.some(c => c.includes('---')));
  if (dataRows.length === 0) return null;
  
  const header = dataRows[0];
  const data = dataRows.slice(1);
  
  return { header, data };
}

// ─── Document builder ─────────────────────────────────────────
async function buildDocument(blocks) {
  const children = [];
  
  // Title page
  children.push(
    new Paragraph({ spacing: { before: 600 } }),
    new Paragraph({
      children: [
        new TextRun({ text: 'TEAM-05 "TurboPapus"', bold: true, size: 52, color: COLORS.primary }),
      ],
      alignment: AlignmentType.CENTER,
      spacing: { after: 100 },
    }),
    new Paragraph({
      children: [
        new TextRun({ text: 'Documentación Completa del Proyecto', size: 36, color: COLORS.secondary }),
      ],
      alignment: AlignmentType.CENTER,
      spacing: { after: 400 },
    }),
    new Paragraph({
      children: [
        new TextRun({ text: 'Plataforma de Inversiones con IA — Módulo de Cobertura Institucional', size: 24, color: '666666' }),
      ],
      alignment: AlignmentType.CENTER,
      spacing: { after: 200 },
    }),
    new Paragraph({
      children: [
        new TextRun({ text: 'Mayo 2026', size: 22, color: '888888' }),
      ],
      alignment: AlignmentType.CENTER,
      spacing: { after: 600 },
    }),
  );
  
  for (const block of blocks) {
    switch (block.type) {
      case 'heading': {
        const headingLevel = block.level;
        let headingSize;
        let headingBefore = 200;
        
        if (headingLevel === 1) { headingSize = 32; headingBefore = 400; }
        else if (headingLevel === 2) { headingSize = 26; headingBefore = 300; }
        else if (headingLevel === 3) { headingSize = 22; headingBefore = 200; }
        else { headingSize = 20; headingBefore = 150; }
        
        children.push(
          new Paragraph({
            children: [
              new TextRun({
                text: block.content,
                bold: true,
                size: headingSize,
                color: COLORS.primary,
              }),
            ],
            spacing: { before: headingBefore, after: 120 },
            border: headingLevel <= 2 ? {
              bottom: { color: COLORS.primary, size: 6, style: BorderStyle.SINGLE, space: 4 },
            } : undefined,
          })
        );
        break;
      }
      
      case 'paragraph': {
        const parts = parseInlineFormatting(block.content);
        children.push(
          new Paragraph({
            children: parts.map(p => new TextRun({
              text: p.text,
              bold: p.bold || false,
              size: 22,
              color: COLORS.text,
            })),
            spacing: { after: 100 },
          })
        );
        break;
      }
      
      case 'list': {
        for (const item of block.items) {
          const parts = parseInlineFormatting(item);
          children.push(
            new Paragraph({
              children: [
                new TextRun({ text: '•  ', size: 22, color: COLORS.primary }),
                ...parts.map(p => new TextRun({
                  text: p.text,
                  bold: p.bold || false,
                  size: 22,
                  color: COLORS.text,
                })),
              ],
              spacing: { after: 60 },
              indent: { left: convertInchesToTwip(0.3) },
            })
          );
        }
        break;
      }
      
      case 'code': {
        const codeLines = block.content.split('\n');
        children.push(
          new Paragraph({
            children: [
              new TextRun({ text: block.lang ? `  ${block.lang}` : '', size: 18, color: '888888', italics: true }),
            ],
            spacing: { before: 120, after: 0 },
            shading: { type: ShadingType.CLEAR, color: COLORS.codeBg },
          })
        );
        
        for (const codeLine of codeLines) {
          children.push(
            new Paragraph({
              children: [
                new TextRun({
                  text: codeLine || ' ',
                  font: 'Consolas',
                  size: 18,
                  color: COLORS.secondary,
                }),
              ],
              spacing: { after: 0, before: 0 },
              indent: { left: convertInchesToTwip(0.2) },
              shading: { type: ShadingType.CLEAR, color: COLORS.codeBg },
            })
          );
        }
        
        children.push(new Paragraph({ spacing: { after: 120 } }));
        break;
      }
      
      case 'table': {
        const parsed = parseTable(block.lines);
        if (parsed) {
          const colWidths = parsed.header.map(() => WidthType.AUTO);
          
          const tableRows = [];
          
          // Header row
          tableRows.push(
            new TableRow({
              tableHeader: true,
              children: parsed.header.map(h => new TableCell({
                children: [
                  new Paragraph({
                    children: [new TextRun({ text: h, bold: true, size: 20, color: COLORS.tableHeaderText })],
                  }),
                ],
                shading: { type: ShadingType.CLEAR, color: COLORS.tableHeader },
                width: { size: 100 / parsed.header.length, type: WidthType.PERCENTAGE },
              })),
            })
          );
          
          // Data rows
          for (const rowData of parsed.data) {
            while (rowData.length < parsed.header.length) rowData.push('');
            tableRows.push(
              new TableRow({
                children: rowData.map((cell, ci) => new TableCell({
                  children: [
                    new Paragraph({
                      children: [new TextRun({ text: cell, size: 20, color: COLORS.text })],
                    }),
                  ],
                })),
              })
            );
          }
          
          children.push(
            new Table({
              rows: tableRows,
              width: { size: 100, type: WidthType.PERCENTAGE },
            }),
            new Paragraph({ spacing: { after: 120 } })
          );
        }
        break;
      }
      
      case 'hr': {
        children.push(
          new Paragraph({
            children: [new TextRun({ text: '____________________________________________________________', size: 16, color: 'CCCCCC' })],
            spacing: { before: 200, after: 200 },
          })
        );
        break;
      }
      
      case 'spacer': {
        children.push(new Paragraph({ spacing: { after: 60 } }));
        break;
      }
    }
  }
  
  return children;
}

// ─── Main ─────────────────────────────────────────────────────
async function main() {
  console.log('📖 Leyendo documentación markdown...');
  const mdContent = fs.readFileSync(MD_PATH, 'utf-8');
  console.log(`   ${(mdContent.length / 1024).toFixed(0)} KB leídos`);
  
  console.log('🔍 Parseando markdown...');
  const blocks = parseMarkdown(mdContent);
  console.log(`   ${blocks.length} bloques identificados`);
  
  console.log('📄 Generando documento Word...');
  const children = await buildDocument(blocks);
  
  const doc = new Document({
    title: 'TEAM-05 TurboPapus - Documentación Completa',
    description: 'Documentación completa del módulo de cobertura institucional',
    styles: {
      default: {
        document: {
          run: { font: 'Calibri', size: 22 },
        },
      },
    },
    sections: [{
      properties: {
        page: {
          margin: {
            top: convertInchesToTwip(0.8),
            right: convertInchesToTwip(0.8),
            bottom: convertInchesToTwip(0.8),
            left: convertInchesToTwip(0.8),
          },
        },
      },
      children,
    }],
  });
  
  console.log('💾 Empaquetando documento...');
  const buffer = await Packer.toBuffer(doc);
  
  console.log(`📁 Escribiendo archivo: ${OUTPUT_PATH}`);
  fs.writeFileSync(OUTPUT_PATH, buffer);
  
  const sizeKB = (buffer.length / 1024).toFixed(1);
  console.log(`✅ Documento generado exitosamente!`);
  console.log(`   Ruta: ${OUTPUT_PATH}`);
  console.log(`   Tamaño: ${sizeKB} KB`);
}

main().catch(err => {
  console.error('❌ Error:', err.message);
  process.exit(1);
});
